kong/plugins/ai-proxy/handler.lua

```diff
@@ -0,0 +1,183 @@
+    this = {
+      messages = {
+        [1] = {
+          role = "system",
+          content = "You are a helpful assistant"
+        },
+        [2] = {
+          role = "user",
+          content = transformation_instructions[1]
+        }
+      }
+    }
+  end
+
+  return this
+end
+
+function _M:header_filter(conf)
+  kong.log.debug("IN: ai-proxy/header_filter")
+
+  -- clear shared restricted headers
+  for i, v in ipairs(ai_shared.clear_response_headers.shared) do
+    kong.response.clear_header(v)
+  end
+
+  -- only act on 200 in first release - pass the unmodifed response all the way through if any failure
+  if kong.response.get_status() == 200 then
+    local content_type = kong.response.get_header("content-type") or "application/json"
+    if content_type then
+      local ai_driver = require("kong.plugins.ai-proxy.drivers." .. conf.model.provider)
+      local route_type = conf.route_type
+      
+      local response_body = kong.service.response.get_raw_body()
+
+      if response_body then
+        local is_gzip = kong.response.get_header("Content-Encoding") == "gzip"
+        if is_gzip then
+          response_body = kong_utils.inflate_gzip(response_body)
+        end
+
+        local new_response_string, err = ai_driver.from_format(response_body, conf.model, route_type)
+        if err then
+          ngx.status = 500
+          kong.ctx.plugin.parsed_response = fmt(INTERNAL_SERVER_ERROR_TEMPLATE, err)
+        end
+        if new_response_string then
+          -- preserve the same response content type; assume the from_format function
+          -- has returned the body in the appropriate response output format
+          kong.ctx.plugin.parsed_response = new_response_string
+        end
+
+        -- else don't transform it
+      end
+    end
+  end
+
+  kong.log.debug("OUT: ai-proxy/header_filter")
+end
+
+function _M:body_filter(conf)
+  kong.log.debug("IN: ai-proxy/body_filter")
+
+  -- all errors MUST be checked and returned in header_filter
+  -- we should receive a replacement response body from the same thread
+  if kong.ctx.plugin.parsed_response then
+    local is_gzip = kong.response.get_header("Content-Encoding") == "gzip"
+    if is_gzip then
+      kong.ctx.plugin.parsed_response = kong_utils.deflate_gzip(kong.ctx.plugin.parsed_response)
+    end
+    kong.response.set_raw_body(kong.ctx.plugin.parsed_response)
+  end
+
+  -- call with replacement body, or original body if nothing changed
+  ai_shared.post_request(conf, kong.ctx.plugin.parsed_response or kong.response.get_raw_body())
+
+  kong.log.debug("OUT: ai-proxy/body_filter")
+end
+
+function _M:access(conf)
+  kong.log.debug("IN: ai-proxy")
+  kong.service.request.enable_buffering()
+
+  -- store the route_type in ctx for use in response parsing
+  local route_type = conf.route_type
+  kong.ctx.plugin.operation = route_type
+
+  if kong.ctx.shared.prompt_guarded and not kong.ctx.shared.prompt_rendered then
```

Comment:
seems kong.ctx.shared.prompt_guarded is not set in any provider

Reply:
This is no longer used, removed it.